src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -0,0 +1,1876 @@
+    		MultidimArray<double> &half1, MultidimArray<double> &half2, bool test)
+    {
+    	MultidimArray<double> &phalf1 = half1, &phalf2 = half2;
+
+    	Image<double> mask;
+    	MultidimArray<double> &pmask = mask();
+
+    	if (test)
+		{
+			Monogenic mono;
+			std::cout << "Preparing test data ..." << std::endl;
+			size_t xdim = 256, ydim = 256, zdim = 256;
+			double wavelength = 5.0, mean = 0.0, std = 0.5;
+			int maskrad = 125;
+			half1 = mono.createDataTest(xdim, ydim, zdim, wavelength, mean, 0.0);
+			half2 = half1;
+
+			mono.addNoise(phalf1, 0, std);
+			mono.addNoise(phalf2, 0, std);
+			FileName fn;
+			Image<double> saveImg;
+			fn = formatString("inputVol1_large.vol");
+			saveImg() = half1;
+			saveImg.write(fn);
+			fn = formatString("inputVol2_large.vol");
+			saveImg() = half2;
+			saveImg.write(fn);
+		}
+		else
+		{
+			std::cout << "Reading data..." << std::endl;
+			Image<double> imgHalf1, imgHalf2;
+			imgHalf1.read(fnhalf1);
+			imgHalf2.read(fnhalf2);
+
+			half1 = imgHalf1();
+			half2 = imgHalf2();
+
+			if (fnmask!="")
+			{
+				mask.read(fnmask);
+				FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(pmask)
+				{
+					double valmask = (double) DIRECT_MULTIDIM_ELEM(pmask, n);
+					DIRECT_MULTIDIM_ELEM(phalf1, n) = DIRECT_MULTIDIM_ELEM(phalf1, n) * valmask;
+					DIRECT_MULTIDIM_ELEM(phalf2, n) = DIRECT_MULTIDIM_ELEM(phalf2, n) * valmask;
+				}
+			}
+			mask.clear();
+			pmask.clear();
+		}
+
+		phalf1.setXmippOrigin();
+		phalf2.setXmippOrigin();
+
+    	std::cout << "Starting..." << std::endl;
+    }
+
+
+    void saveFSCToMetadata(MetaData &mdRes,
+    		const MultidimArray<double> &freq,
+			const MultidimArray<double> &FSC, FileName &fnmd)
+    {
+    	size_t id;
+    	FOR_ALL_ELEMENTS_IN_ARRAY1D(freq)
+		{
+			if (i>0)
+			{
+				id=mdRes.addObject();
+				mdRes.setValue(MDL_RESOLUTION_FREQ,dAi(freq, i),id);
+				mdRes.setValue(MDL_RESOLUTION_FRC,dAi(FSC, i),id);
+				mdRes.setValue(MDL_RESOLUTION_FREQREAL,1./dAi(freq, i),id);
+			}
+		}
+		mdRes.write(fnmd);
+    }
+
+    void saveAnisotropyToMetadata(MetaData &mdAnisotropy,
+    		const MultidimArray<double> &freq,
+			const MultidimArray<double> &anisotropy, FileName &fnmd)
+    {
+    	size_t objId;
+		FOR_ALL_ELEMENTS_IN_ARRAY1D(anisotropy)
+		{
+			if (i>0)
+			{
+			objId = mdAnisotropy.addObject();
+			mdAnisotropy.setValue(MDL_RESOLUTION_FREQ, dAi(freq, i),objId);
+			mdAnisotropy.setValue(MDL_RESOLUTION_FRC, dAi(anisotropy, i),objId);
+			mdAnisotropy.setValue(MDL_RESOLUTION_FREQREAL, 1.0/dAi(freq, i),objId);
+			}
+		}
+		mdAnisotropy.write(fnmd);
+    }
+
+
+
+    void directionalFilter(MultidimArray<std::complex<double>> &FThalf1,
+    		MultidimArray<double> &threeDfsc, MultidimArray<double> &filteredMap, int m1sizeX, int m1sizeY, int m1sizeZ)
+    {
+
+    	Image<double> imgHalf1;
+    	imgHalf1.read(fnhalf1);
+    	MultidimArray<double> half1;
+    	half1 = imgHalf1();
+
+        FourierTransformer transformer1(FFTW_BACKWARD);
```

Comment:
missing multithreading?
add fixme to use `atf.h`

Reply:
It is just only one transform, I prefer to avoid problems, and the method is fast enough
The FIXME is added